### PR TITLE
Restructure

### PR DESCRIPTION
**Pull request by Hoverbear**
Builds off #46.

Restructures how we do crate generation.

We now have a state machine `UserCrate` which has three states:

* `StateGenerated`: Constructed from a fn_oid or components this can output `lib.rs` or `Cargo.toml`s and can state change via `provision()` into a `StateProvisioned`
* `StateProvisioned`: Resident to disk, able to state change via `build()` to `StateBuilt`
* `StateBuilt`: A shared object now exists in the workdir, holds that path.
* `StateLoaded`: A  loaded shared object (with a late initialized symbol)

Adds some testing around the code.

No significant changes to parsing/generation were done, it was mostly movement.

### CODE REVIEW

**Comment by Hoverbear** [comment]
The only commented code here is a little `TODO` idea for me (`Trigger`) and a test introduced in a separate PR (#46 ).

Also this is a draft and not ready for review.

**Comment by Hoverbear** [comment]
Alright I'm happy with this now.

File: src/lib.rs [hunk 1]
```diff
@@ -118,6 +122,8 @@ unsafe fn plrust_validator(fn_oid: pg_sys::Oid, fcinfo: pg_sys::FunctionCallInfo
         }
 
         // if the compilation had warnings we'll display them
+        let stderr =
+            String::from_utf8(output.stdout.clone()).expect("`cargo`'s stdout was not  UTF-8");
```
Comment (workingjubilee) [inline src/lib.rs:126]:
```suggestion
            String::from_utf8(output.stdout.clone()).expect("`cargo`'s stdout was not UTF-8");
```

File: src/lib.rs [hunk 2]
```diff
@@ -146,10 +152,14 @@ fn recompile_function(
...
+        Ok((work_dir, output)) => (
             Some(work_dir.display().to_string()),
-            Some(stdout),
-            Some(stderr),
+            Some(
+                String::from_utf8(output.stdout.clone()).expect("`cargo`'s stdout was not  UTF-8"),
+            ),
+            Some(
+                String::from_utf8(output.stderr.clone()).expect("`cargo`'s stderr was not  UTF-8"),
+            ),
```
Comment (workingjubilee) [inline src/lib.rs:162]:
```suggestion
                String::from_utf8(output.stdout.clone()).expect("`cargo`'s stdout was not UTF-8"),
            ),
            Some(
                String::from_utf8(output.stderr.clone()).expect("`cargo`'s stderr was not UTF-8"),
            ),
```

File: src/user_crate/mod.rs [hunk 3]
```diff
@@ -0,0 +1,333 @@
...
+
+            let parent_dir = tempdir::TempDir::new("plrust-generated-crate-function-workflow")
+                .wrap_err("Creating temp dir")?;
+            let provisioned = generated.provision(parent_dir.path())?;
+
+            let (built, _output) =
+                provisioned.build(parent_dir.path(), pg_config, Some(target_dir.as_path()))?;
+
+            let _shared_object = built.shared_object();
+            // TODO: Assert
```
Comment (workingjubilee) [inline src/user_crate/mod.rs:324]: Hm? What is this TODO waiting on, exactly? Assert what, the fcinfo condition?

**Review by workingjubilee — APPROVED** [review]
Looks good modulo trivial nits.

File: src/user_crate/mod.rs (hunk 3 above)
Comment (Hoverbear) [reply src/user_crate/mod.rs:324]: Uhhhh rereading it, I guess I wanted to assert a given symbol existed.

**Comment by Hoverbear** [comment]
Will merge after CI

**Pull request merged**